File: GLLara/GLLItemMeshTexture.m
//
//  GLLItemMeshTexture.m
//  GLLara
//
//  Created by Torsten Kammer on 03.02.13.
//  Copyright (c) 2013 Torsten Kammer. All rights reserved.
//

#import "GLLItemMeshTexture.h"

#import "GLLItemMesh.h"

#import "GLLara-Swift.h"

@implementation GLLItemMeshTexture

@dynamic textureURL;
@dynamic textureBookmarkData;
@dynamic identifier;
@dynamic texCoordSet;
@dynamic mesh;

- (void)awakeFromFetch
{
    [super awakeFromFetch];
    
    // Get URL from bookmark
    NSData *bookmarkData = self.textureBookmarkData;
    if (bookmarkData)
    {
        NSURL *textureURL = [NSURL URLByResolvingBookmarkData:bookmarkData options:0 relativeToURL:nil bookmarkDataIsStale:NULL error:NULL];
        [self setPrimitiveValue:textureURL forKey:@"textureURL"];
    }
}

- (void)willSave
{
    // Put URL into bookmark
    NSURL *textureURL = [self primitiveValueForKey:@"textureURL"];
    if (textureURL)
    {
        NSData *bookmark = [textureURL bookmarkDataWithOptions:0 includingResourceValuesForKeys:nil relativeToURL:nil error:NULL];
        [self setPrimitiveValue:bookmark forKey:@"textureBookmarkData"];
    }
    else
        [self setPrimitiveValue:nil forKey:@"textureBookmarkData"];
}

- (void)setTexCoordSet:(int16_t)texCoordSet {
    [self willChangeValueForKey:@"texCoordSet"];
    [self setPrimitiveValue:@(texCoordSet) forKey:@"texCoordSet"];
    [self.mesh updateShader];
    [self didChangeValueForKey:@"texCoordSet"];
}

- (GLLTextureDescription *)textureDescription
{
    return [self.mesh.mesh.shader descriptionForTexture:self.identifier];
}

@end

File: GLLara/GLLItemMeshTexture.swift
//
//  GLLItemMeshTexture+CoreDataClass.swift
//  GLLara
//
//  Created by Torsten Kammer on 03.05.24.
//  Copyright © 2024 Torsten Kammer. All rights reserved.
//
//

import Foundation
import CoreData

@objc(GLLItemMeshTexture)
public class GLLItemMeshTexture: NSManagedObject {
    
    // Get URL from bookmark
    @objc public override func awakeFromFetch() {
        if let data = textureBookmarkData {
            var stale = false
            if let textureUrl = try? URL(resolvingBookmarkData: data, options: [], relativeTo: nil, bookmarkDataIsStale: &stale) {
                self.setPrimitiveValue(textureUrl, forKey: "textureURL")
            }
        }
    }
    
    @objc public override func willSave() {
        // Put URL into bookmark
        if let url = self.primitiveValue(forKey: "textureURL") as? NSURL {
            let textureUrl = url as URL
            if let bookmark = try? textureUrl.bookmarkData() {
                self.setPrimitiveValue(bookmark, forKey: "textureBookmarkData")
            }
        } else {
            self.setPrimitiveValue(nil, forKey: "textureBookmarkData")
        }
    }
    
    @objc var texCoordSet: Int {
        set {
            willChangeValue(for: \.texCoordSet)
            setPrimitiveValue(NSNumber(value: newValue), forKey: "texCoordSet")
            mesh?.updateShader()
            didChangeValue(for: \.texCoordSet)
        }
        get {
            return (primitiveValue(forKey: "texCoordSet") as! NSNumber).intValue
        }
    }
    
    @objc var textureDescription: GLLTextureDescription? {
        guard let identifier, let mesh, let shader = mesh.mesh.shader else {
            return nil
        }
        return shader.description(forTexture: identifier)
    }
}

extension GLLItemMeshTexture {

    @nonobjc public class func fetchRequest() -> NSFetchRequest<GLLItemMeshTexture> {
        return NSFetchRequest<GLLItemMeshTexture>(entityName: "GLLItemMeshTexture")
    }

    @NSManaged public var identifier: String?
    @NSManaged public var textureBookmarkData: Data?
    @NSManaged public var textureURL: NSURL?
    @NSManaged public var mesh: GLLItemMesh?
}
